packages/react-juce/src/lib/MacroProperties/Transform.ts

```diff
@@ -0,0 +1,130 @@
+import { TPropertyAssignment } from "./types";
+import { getMacroCalls } from "./util";
+import matrix from 'matrix-js';
+
+const rotateMultipliers = {
+  deg: Math.PI / 180.0,
+  grad: Math.PI / 200.0,
+  rad: 1,
+  turn: 2 * Math.PI,
+};
+
+const angleUnitMatcher = new RegExp(
+  `(-?[0-9]*\.?[0-9]+)(${Object.keys(rotateMultipliers).join("|")}|)$`
+);
+
+const toRadians = (arg: string) => {
+  // @ts-ignore
+  let [, angle, unit] = angleUnitMatcher.exec(arg);
+  let angleFloat = parseFloat(angle);
+  if (angleFloat === NaN) return NaN;
+  if (unit === "" && angleFloat !== 0) return NaN;
+  angleFloat *= rotateMultipliers[unit] || 1;
+  return angleFloat;
+}
+
+const argsToFloat = f => (...args) => f(...args.map(parseFloat));
+
+const identity = matrix([
+  [1, 0, 0, 0],
+  [0, 1, 0, 0],
+  [0, 0, 1, 0],
+  [0, 0, 0, 1],
+]);
+
+const rotate3d = (x, y, z, theta) => {
+  const t = toRadians(theta) || 0;
+  const sinTheta = Math.sin(t);
+  const cosTheta = Math.cos(t);
+  const [floatX, floatY, floatZ] = [x, y, z].map(parseFloat);
+  const r = Math.sqrt(floatX ** 2 + floatY ** 2 + floatZ ** 2)
+  if (r === 0) { return identity; }
+  const [u, v, w] = [floatX / r, floatY / r, floatZ / r];
+
+  return matrix([
+    [cosTheta + (1 - cosTheta) * u ** 2, u * v * (1 - cosTheta) - w * sinTheta, u * w * (1 - cosTheta) + v * sinTheta, 0],
+    [v * u * (1 - cosTheta) + w * sinTheta, cosTheta + (1 - cosTheta) * v ** 2, v * w * (1 - cosTheta) - u * sinTheta, 0],
+    [w * u * (1 - cosTheta) - v * sinTheta, w * v * (1 - cosTheta) + u * sinTheta, cosTheta + (1 - cosTheta) * w ** 2, 0],
+    [0, 0, 0, 1]
+  ]);
+}
+
+const rotateX = (theta) => rotate3d(1, 0, 0, theta);
+const rotateY = (theta) => rotate3d(0, 1, 0, theta);
+const rotateZ = (theta) => rotate3d(0, 0, 1, theta);
+
+const scale3d = (sx = 1, sy = 1, sz = 1) => matrix([
+  [sx, 0, 0, 0],
+  [0, sy, 0, 0],
+  [0, 0, sz, 0],
+  [0, 0, 0, 1]
+]);
+
+const scale = argsToFloat((sx = 1, sy = 1) => scale3d(sx, sy));
+const scaleX = argsToFloat((sx = 1) => scale(sx));
+const scaleY = argsToFloat((sy = 1) => scale(1, sy));
+const scaleZ = argsToFloat((sz = 1) => scale3d(1, 1, sz));
+
+const skew = (alpha = '0', beta = '0') => {
+  // the || 0 is lazy coding, we can address clients
+  // who want to skew an axis by 90deg later
+  const tanAlpha = Math.tan(toRadians(alpha) || 0)
+  const tanBeta = Math.tan(toRadians(beta) || 0)
+  return matrix([
+    [1 + tanAlpha * tanBeta, tanAlpha, 0, 0],
+    [tanBeta, 1, 0, 0],
+    [0, 0, 1, 0],
+    [0, 0, 0, 1],
+  ]);
+}
+
+const skewX = (theta) => skew(theta);
+const skewY = (theta) => skew('0', theta);
+
+const translate3d = argsToFloat((dx = 0, dy = 0, dz = 0) => matrix([
```

Comment:
Not sure I understand the `argsToFloat` part here either.

I think it'd be easier to understand if all of our rotate/scale/translate/etc helpers operated over strictly numeric values of a consistent unit (i.e. radians). Any conversions that we need to do to go from the user's string argument `rotate(90deg)` to that numeric value can be done above this layer.

The other reason I think it's helpful to think that way is that `const argsToFloat = f => (...args) => f(...args.map(parseFloat));` is quite an expensive wrapper :). Most major ES6->ES5 transpilers will rewrite that touching the `arguments` object, which is a known performance in Duktape and often a deopt flag in V8/SpiderMonkey. So, liberally wrapping our helper functions in this could end up being quite costly for apps that use these `transform` properties regularly

Reply:
I see what you mean about the mathematical functions being relieved of concern about strings.  The part about the expensive wrapper is a bit above my paygrade, so I'd appreciate your thoughts on what I've done.  I moved the string argument to float translation functions out to a map, and compose those with purely numerical args => matrix functions.  Is this actually an improvement at all in term of the transpiled code?

Reply:
Yea, it's tricky :)

If we're running in an engine like V8, this is definitely an improvement in the transpiled code, because we're touching the `arguments` object in fewer places (which means fewer de-optimization flags).

In Duktape, we still will pay the penalty for touching that `arguments` object in those 2 places, since each transform function has to first run through it string->float/radian conversion first, but again, less touching `arguments` = better.

So to me this looks good! I think this is about as good as we're going to get without another very long, very detailed performance pass. And if we do that performance pass, we shouldn't block the feature from merging before we do.